Migrating PHPUnit tests to Kahlan tests.

william-man opened the pull request:
# Description

This pull request relates to issue #17 to migrate old PHPUnit tests to Kahlan so that only one testing library is needed to learn in detail.

## Changes

- fixed typo for layout_register_test.php to make sure the all original tests run and succeeds.
- added Kahlan --v 6.0 to run the new Kahlan tests.
- created the 3 new Kahlan tests into the spec folder. To run the tests use ```./vendor/bin/kahlan``` or individually with ```./vendor/bin/kahlan --spec=spec/test_name.spec.php```

## Problems

- An error occurs when running the new ```layout_register_test.spec.php``` as shown below
```
an uncaught exception has been thrown in `src/Testing.php` line 14
message:`Error` Code(0) with message "Call to undefined method LayoutRegisterTestHelper::getMockBuilder()"
```
which seems to be coming from the Testing.php file in the src folder: https://github.com/dxw/iguana-theme/blob/main/src/Testing.php . It might involve the getMockBuilder at line 14 which I found to be something PHPunit uses to mock an object as described in the docs here: https://docs.phpunit.de/en/9.6/test-doubles.html?highlight=getmockbuilder.

snim2 commented on tests/layout_register_test.php:
```diff
@@ -1,6 +1,6 @@
 <?php
 
-class LayoutRegister_Test extends \PHPUnit\Framework\TestCase
+class Layout_Register_Test extends \PHPUnit\Framework\TestCase
```
This wasn't a typo, the unit under test is `LayoutRegister` and the test class needs to end in `_Test` to meet the phpunit spec

william-man replied:
Strange. When I run ```vendor/bin/phpunit tests/layout_register_test.php``` it gives me an error:
```
Class layout_register_test could not be found  in .../iguana-theme/tests/layout_register_test.php
```
but it is fine if I add the underscore.

snim2 replied:
From your `main` branch if I run `composer install` I get this:

```shell
❯ ./vendor/bin/phpunit
PHPUnit 9.6.22 by Sebastian Bergmann and contributors.

........I                                                           9 / 9 (100%)

Time: 00:00.029, Memory: 8.00 MB

OK, but incomplete, skipped, or risky tests!
Tests: 9, Assertions: 14, Incomplete: 1.
```

On both PHP v7.4 and 8.3

william-man replied:
ok I've just done the same on both PHP v7.4 and 8.3 and it shows the same as you've posted. Shall I revert that change to the name of the class?

william-man commented:
## Update

Updated the specs to follow the example kahlan test provided. Removed WP_Mock and instead used kahlan's allow() to stub functions/methods.  Fixed the errors when running layout_register_test to not use the Testing class.

snim2 commented:
@william-man next steps for this one:

* remove the WP_Mock and phpunit dependencies from Composer and upgrade
* change the GitHub Action to run the new tests
* make sure composer installs correctly you've added Kahlan to `composer.json` but not committed the changes to the `.lock` file that come after running `composer upgrade`. Delete `composer.lock` because the convention in PHP is not to commit lockfiles for libraries
* update the Changelog, you can see the format for this file [in the documentation here](https://keepachangelog.com/en/1.0.0/)
* Read about rebasing in Git:
   * [in this tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase)
   * [and this video](https://www.youtube.com/watch?v=f1wnYdLEpgI)
* Re-write the PR description as per [our docs](https://playbook.dxw.com/govpress-unit/code-review/) (e.g. see [this example](https://github.com/dxw/whippet/pull/212))

snim2 commented on src/Testing.php:
```diff
@@ -1,40 +0,0 @@
-<?php
```
@RobjS For this PR William is migrating the unit tests to Kahlan and I /think/ this trait is something we no longer need once we move the other repos away from phpunit, so it's OK to delete and make a major version release.

Is that right or should we be doing something else here?

RobjS replied:
I think that's right. To the best of my knowledge, this trait isn't in use anywhere else. And if it is, the worst that should happen is that the unit tests fail where this is in use, rather than any production code being affected.

snim2 commented on CHANGELOG.md:
```diff
@@ -5,6 +5,22 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [v1.2.0] - 2025-01-17
```
Let's say `[Unreleased]` here

snim2 commented on src/Testing.php:
```diff
@@ -1,40 +0,0 @@
-
-namespace Dxw\Iguana\Theme;
-
-trait Testing
```
@william-man let's reword this commit message to say why we're deleting this (because we no longer use PHPUnit so it's not useful to us)

snim2 commented on spec/helpers_test.spec.php:
```diff
@@ -0,0 +1,31 @@
+	beforeEach(function () {
+		$this->helpers = new \Dxw\Iguana\Theme\Helpers();
+	});
+	afterEach(function () {
```
We can remove this because it's empty

snim2 commented on spec/helpers_test.spec.php:
```diff
@@ -0,0 +1,31 @@
+	it("test register", function () {
+		expect($this->helpers)->toBeAnInstanceOf(\Dxw\Iguana\Registerable::class);
+
+		expect(function_exists('h'))->toBeFalsy();
```
I think this should be `toBeFalse()`

snim2 commented on spec/helpers_test.spec.php:
```diff
@@ -0,0 +1,31 @@
+
+		expect(function_exists('h'))->toBeFalsy();
+		$this->helpers->register();
+		expect(function_exists('h'))->toBeTruthy();
```
`toBeTrue()`

snim2 commented on spec/layout_register_test.spec.php:
```diff
@@ -0,0 +1,40 @@
+<?php
+
+describe(\Dxw\Iguana\Theme\LayoutRegister::class, function () {
+	beforeEach(function () {
```
I don't think this stuff with the template file should be in the setup/teardown, it's a separate test

snim2 commented on spec/layout_register_test.spec.php:
```diff
@@ -0,0 +1,40 @@
+
+	it("test constructor", function () {
+		$this->helpers->__call('w_requested_template', []);
+		global $called;
```
I don't think the original test here checked `called`

snim2 commented on spec/layout_register_test.spec.php:
```diff
@@ -0,0 +1,40 @@
+
+	it("test wRequestedTemplate", function () {
+		global $called;
+		$this->layoutRegister->wRequestedTemplate();
```
This is where the file creation needs to happen

snim2 commented on spec/layout_test.spec.php:
```diff
@@ -0,0 +1,39 @@
+	});
+
+	afterEach(function () {
+		\Dxw\Iguana\Theme\Layout::$base = null;
```
The original test also had:

```php
		\Dxw\Iguana\Theme\Layout::$wordpress_template = null;
```

snim2 commented on spec/layout_test.spec.php:
```diff
@@ -0,0 +1,39 @@
+		expect($this->layout->__toString())->toBe('correct output');
+	});
+
+	it("test constructor", function () {
```
The original test was:

```php
public function testConstructor()
	{
		$this->markTestIncomplete();
	}
```

although it would be good to make a better test!

The code under test here would be this:

```php
public function __construct($template = 'layouts/main.php')
	{
		$this->slug = basename($template, '.php');
		$this->templates = [$template];

		if (self::$base) {
			$str = substr($template, 0, -4);
			array_unshift($this->templates, sprintf($str.'-%s.php', self::$base));
		}
	}
```

snim2 commented on spec/layout_test.spec.php:
```diff
@@ -0,0 +1,39 @@
+
+	it("test constructor", function () {
+		expect($this->layout->slug)->toBe('main');
+		expect($this->layout->templates)->toBe(['layouts/main.php']);
```
I think if the input here is `layouts/main.php` then we're expecting `main`?

snim2 replied:
(for `$slug`)

snim2 replied:
and then if `self::$base` is set then we're adding another item to the `$templates` array

william-man commented on spec/layout_register_test.spec.php:
```diff
@@ -0,0 +1,42 @@
+		expect($value)->toBe(1);
+		unlink('mock_template.php');
+	});
+
```
I'm not sure if this is the right way to test the constructor because I think the original test used PHPUnit to mock the Helpers class, so instead I decided to check if wRequestedTemplate is called if it is registered with the Helpers' register function.

snim2 replied:
Did the original test write out a file and delete it?

william-man replied:
The original test used PHPUnit to mock the Helpers class, create a new LayoutRegister object with the mock and then checks whether the w_requested_template function is called. But I thought, since I've already initialised a Helpers object, it'll be easier to just check w_requested_template is registered by creating a mock file and check to see if it is required then delete it once the test is complete.

snim2 replied:
So my question was about whether it created a file, the answer is here: https://github.com/dxw/iguana-theme/blob/main/tests/layout_register_test.php#L45, so it did create a file but it used a mock filesystem.

I'm not sure then whether you need to `unlink()` the file later? Do you see it on disk if you remove that line?

william-man replied:
Oh yes, the original test for `wRequestedTemplate` method does create a file so I took a similar approach with a separate file and moved what I wrote in the `afterEach()`back into the wRequestedTemplate test. 

The new file for the constructor does remain if `unlink()` is removed.
